lms/djangoapps/instructor/views/api.py

```diff
@@ -2910,3 +2910,98 @@ def generate_certificate_exceptions(request, course_id, generate_for=None):
+                                  'extraneous characters or rows.')
+                })
+
+        except Exception:  # pylint: disable=broad-except
+            errors.append({
+                'response': _('-- Could not read uploaded file.')
+            })
+        finally:
+            upload_file.close()
+
+        row_num = 0
+        for student in students:
+            row_num += 1
+            # verify that we have exactly two column in every row either email or username and notes but allow for
+            # blank lines
+            if len(student) != 2:
+                if len(student) > 0:
+                    errors.append({
+                        'response': _('-- Data in row# {num} must have exactly two columns, '
+                                      'either an email or a username and notes.').format(num=row_num)
+                    })
+                    log.info(u'invalid data/format in csv row# %s', row_num)
+                continue
+            user = student[USER_INDEX]
+            try:
+                db_user = get_user_by_username_or_email(user)
+            except ObjectDoesNotExist:
+                errors.append({'response': _('-- user (username/email={user}) does not exist.').format(user=user)})
+                log.info(u'student %s does not exist', user)
+            else:
+                if len(CertificateWhitelist.get_certificate_white_list(course_key, db_user)) > 0:
+                    warnings.append({
+                        'response': _("-- user '{username}' in row# {row} is already in certificate exceptions "
+                                      "list.").format(username=db_user.username, row=row_num)
+                    })
+                    log.warning(u'student %s already exist.', db_user.username)
+
+                # make sure user is enrolled in course
+                elif not CourseEnrollment.is_enrolled(db_user, course_key):
+                    warnings.append({
+                        'response': _("-- user '{username}' in row# {row} is not enrolled in course.").format(
+                            username=db_user.username, row=row_num
+                        )
+                    })
+                    log.warning(u'student %s is not enrolled in course.', db_user.username)
+
+                else:
+                    CertificateWhitelist.objects.create(
+                        user=db_user,
+                        course_id=course_key,
+                        whitelist=True,
+                        notes=student[NOTES_INDEX]
+                    )
+                    success.append({
+                        'response': _("-- user '{username}' successfully added to exception list.").format(
+                            username=db_user.username
+                        )
+                    })
+
+    else:
+        errors.append({
+            'response': _('-- File is not attached.')
+        })
+
+    results = {
```

Comment:
I'm concerned with amount of data this API returns. If we we upload a CSV with 2000 rows this api would return 2000 rows having strings like `-- user {username} successfully added to exception list.` and that consume some bandwidth and as a result of it user experience would be slow. What if we return only list of username names and show different message for `success`, `error` and `warnings`.